crates/lang/codegen/src/generator/dispatch.rs

```diff
@@ -740,24 +739,33 @@ impl Dispatch<'_> {
 
             quote_spanned!(message_span=>
                 Self::#message_ident(input) => {
-                    let config = ::ink_lang::codegen::ExecuteMessageConfig {
-                        payable: #accepts_payment,
-                        mutates: #mutates_storage,
-                    };
-                    let mut contract: ::core::mem::ManuallyDrop<#storage_ident> =
-                        ::core::mem::ManuallyDrop::new(
-                            ::ink_lang::codegen::initiate_message::<#storage_ident>(config)?
-                        );
+                    use ::core::default::Default;
+
+                    if #deny_payment {
+                        ::ink_lang::codegen::deny_payment::<
+                            <#storage_ident as ::ink_lang::reflect::ContractEnv>::Env>()?;
+                    }
+
                     let result: #message_output = #message_callable(&mut contract, input);
                     let failure = ::ink_lang::is_result_type!(#message_output)
                         && ::ink_lang::is_result_err!(result);
-                    ::ink_lang::codegen::finalize_message::<#storage_ident, #message_output>(
-                        !failure,
-                        &contract,
-                        config,
-                        &result,
-                    )?;
-                    ::core::result::Result::Ok(())
+
+                    if failure {
+                        // There is no need to push back the intermediate results of the
+                        // contract since the transaction is going to be reverted.
```

Comment:
```suggestion
                        // We return early here since there is no need to push back the
                        // intermediate results of the contract - the transaction is going to be
                        // reverted anyways.
```